fms/utils/serialization.py

```diff
@@ -138,12 +226,257 @@ def load_state_dict(
 
         checkpoints = [checkpoints[rank]]
 
-    checkpoint_sds = [
-        torch.load(ckpt_path, map_location="cpu") for ckpt_path in checkpoints
-    ]
-    assert len(checkpoint_sds[0]), f"Unable to load checkpoint data at {model_path}"
-    if len(checkpoint_sds) == 1:
-        return checkpoint_sds[0]
+    checkpoint_sds = []
+    if checkpoint_format == "st":
+        for ckp in checkpoints:
+            checkpoint_sds.append(
+                load_safetensors_state_dict(
+                    ckp,
+                    initial_device,
+                )
+            )
     else:
-        # layer-sharded checkpoints, e.g. as used in HF
-        return ChainMap(*checkpoint_sds)
+        with torch.no_grad():
+            checkpoint_sds = [
+                torch.load(str(ckpt_path), mmap=True) for ckpt_path in checkpoints
+            ]
+    return ChainMap(*checkpoint_sds)
+
+
+def load_safetensors_state_dict(
+    checkpoint: Path,
+    device: torch.device,
+):
+    sd = LazySafetensorsDict()
+
+    from safetensors import safe_open  # type: ignore[import-untyped]
+
+    with safe_open(
+        checkpoint, framework="pt", device=str(device)
+    ) as model_weights:  # type: ignore[attr-defined]
+        sd_keys = list(model_weights.keys())
+        for key in sd_keys:
+            sd.set_lazy_tensor(key, checkpoint, device)
+    return sd
+
+
+class FusableWeightsMissingError(Exception):
+    missing_weights: List[str] = []
+
+    def __init__(self, missing_weights):
+        self.missing_weights = missing_weights
+        super().__init__()
+
+
+def load_state_dict_into_model(
+    model: torch.nn.Module,
+    state_dict: MutableMapping[str, Any],
+    architecture: str,
+    source: str,
+    distributed_strategy: Optional[str] = None,
+    checkpoint_sharding: Optional[str] = None,
+    initial_device: torch.device = torch.device("cpu"),
+    rank: int = 0,
+    world_size: int = 0,
+) -> None:
+    # 1. Get the adapter from checkpoint sd to fms sd
+    adapter = _get_adapter(architecture, source)
+
+    # 2. Decide if model needs sharding and how (for now only TP)
+    needs_tp_sharding = checkpoint_sharding != "tp" and distributed_strategy == "tp"
+
+    # 3. Iterate over the weights and load them into the model
+    used_keys = set()
+    sd_keys = list(state_dict.keys())
+    with torch.no_grad():
+        for key in sd_keys:
+            if key in used_keys:
+                continue
+            used_keys.add(key)
+            try:
+                partial_sd = {key: state_dict[key]}
+                if partial_sd[key].device != initial_device:
+                    partial_sd[key] = partial_sd[key].to(device=initial_device)
+                fms_partial_sd = adapter(partial_sd)
+            except FusableWeightsMissingError as e:
```

Comment:
I generally don't love use of exceptions for control flow but this might be our best option. I haven't read in detail yet but it certainly looks cleaner than the api in the previous revision

Reply:
the reasoning for this approach is the following: when you have an adapter that turns N weights into 1, you should already have code that checks that your checkpoint state_dict has all the weights you are trying to fuse. Given you should error out on that, we can "hijack" the exception with our custom one, and have the function return what weights it's missing, in which case the caller can look for the missing weights and add them to the state dict if they exist. If they really don't, we just bubble the error up.